File: Leetcode 3501-3600/3566/3566.py
# Leetcode 3566: Partition Array into Two Equal Product Subsets
# https://leetcode.com/problems/partition-array-into-two-equal-product-subsets/
# Solved on 3rd of December, 2025
class Solution:
    def chckEqualPartitions(self, nums: list[int], target: int) -> bool:
        """
        Checks if the given array `nums` can be partitioned into two subsets such that the product of elements
        in each subset is equal to `target`.

        :param nums: A list of integers.
        :param target: The target product for each subset.
        :return: True if such a partition is possible, False otherwise.
        """
        totalProduct = 1
        for num in nums:
            totalProduct *= num

        if totalProduct != target * target:
            return False

        nums.sort(reverse=True)
        length = len(nums)

        def canPartition(index, currentProduct):
            if currentProduct == target:
                return True
            if currentProduct > target or index == length:
                return False

            if canPartition(index + 1, currentProduct * nums[index]):
                return True

            if canPartition(index + 1, currentProduct):
                return True

            return False

        return canPartition(0, 1)

File: Leetcode 3501-3600/3566/3566.java
// Leetcode 3566: Partition Array into Two Equal product Subsets
// https://leetcode.com/problems/partition-array-into-two-equal-product-subsets/
// Solved on 3rd of December, 2025
class Solution {
    public boolean checkEqualPartitions(int[] nums, long target) {
        /**
         * Checks if the given array `nums` can be partitioned into two subsets such that the product of elements in each subset equals the `target`.
         * @param nums The input array of integers.
         * @param target The target product for each subset.
         * @return True if such a partition is possible, false otherwise.
         */
        int n = nums.length;
        int limit = 1 << (n - 1);
        for (int i = 1; i < limit; i++) {
            long product1 = 1;
            long product2 = 1;
            boolean possible = true;
            for (int j = 0; j < n; j++) {
                if ((i & (1 << j)) != 0) {
                    product1 *= nums[j];
                    if (product1 > target) {
                        possible = false;
                        break;
                    }
                } else {
                    product2 *= nums[j];
                    if (product2 > target) {
                        possible = false;
                        break;
                    }
                }
            }
            if (possible && product1 == target && product2 == target) {
                return true;
            }
        }
        return false;
    }
}